src/cmd/linuxkit/push_aws.go

```diff
@@ -0,0 +1,165 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+const timeoutVar = "LINUXKIT_UPLOAD_TIMEOUT"
+
+// Process the run arguments and execute run
+func pushAWS(args []string) {
+	awsCmd := flag.NewFlagSet("aws", flag.ExitOnError)
+	invoked := filepath.Base(os.Args[0])
+	awsCmd.Usage = func() {
+		fmt.Printf("USAGE: %s push aws [options] [name]\n\n", invoked)
+		fmt.Printf("'name' specifies the full path of an image file which will be uploaded\n")
+		fmt.Printf("Options:\n\n")
+		awsCmd.PrintDefaults()
+	}
+	timeoutFlag := awsCmd.Int("timeout", 0, "Upload timeout in seconds")
+	bucketFlag := awsCmd.String("bucket", "", "S3 Bucket to upload to. *Required*")
+	nameFlag := awsCmd.String("img-name", "", "Overrides the Name used to identify the file in Amazon S3 and Image. Defaults to [name] with the file extension removed.")
+
+	if err := awsCmd.Parse(args); err != nil {
+		log.Fatal("Unable to parse args")
+	}
+
+	remArgs := awsCmd.Args()
+	if len(remArgs) == 0 {
+		fmt.Printf("Please specify the path to the image to push\n")
+		awsCmd.Usage()
+		os.Exit(1)
+	}
+	src := remArgs[0]
+
+	timeout := getIntValue(timeoutVar, *timeoutFlag, 600)
+	bucket := getStringValue(bucketVar, *bucketFlag, "")
+	name := getStringValue(nameVar, *nameFlag, "")
+
+	sess := session.Must(session.NewSession())
+	storage := s3.New(sess)
+
+	ctx, cancelFn := context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
+	defer cancelFn()
+
+	if bucket == "" {
+		log.Fatalf("No bucket specified. Please provide one using the -bucket flag")
+	}
+
+	f, err := os.Open(src)
+	if err != nil {
+		log.Fatalf("Error opening file: %s", err)
+	}
+	defer f.Close()
+
+	if name == "" {
+		name = src[0 : len(src)-len(filepath.Ext(src))]
+	}
+
+	content, err := ioutil.ReadAll(f)
+	if err != nil {
+		log.Fatalf("error reading file: %s", err)
+	}
+
+	putParams := &s3.PutObjectInput{
+		Bucket:        aws.String(bucket),
+		Key:           aws.String(fmt.Sprintf("%s.%s", name, filepath.Ext(src))),
+		Body:          f,
+		ContentLength: aws.Int64(int64(len(content))),
+		ContentType:   aws.String("application/octet-stream"),
+	}
+	log.Debugf("PutObject:\n%v", putParams)
+
+	_, err = storage.PutObjectWithContext(ctx, putParams)
+	if err != nil {
+		log.Fatalf("Error uploading to S3: %s", err)
+	}
+
+	compute := ec2.New(sess)
+
+	importParams := &ec2.ImportSnapshotInput{
+		Description: aws.String(fmt.Sprintf("LinuxKit: %s", name)),
+		DiskContainer: &ec2.SnapshotDiskContainer{
+			Description: aws.String(fmt.Sprintf("LinuxKit: %s disk", name)),
+			Format:      aws.String("raw"),
+			UserBucket: &ec2.UserBucket{
+				S3Bucket: aws.String(bucket),
+				S3Key:    aws.String(src),
```

Comment:
here you should change `src` into the new file name created with `<name>.img`, otherwise the file names differ and you're referencing the wrong S3 file